pt-lua.c

```diff
@@ -0,0 +1,966 @@
+  sa.sa_handler = handler;
+  sa.sa_flags = 0;
+  sigemptyset(&sa.sa_mask);  /* do not mask any signal */
+  sigaction(sig, &sa, NULL);
+}
+
+#else           /* }{ */
+
+#define setsignal            signal
+
+#endif                               /* } */
+
+
+/* Global table name deduction. Can we find a function name? */
+static bool findfield(lua_State *L, int fn_idx, int level) {
+  if(level == 0 || !lua_istable(L, -1))
+    return false;
+
+  lua_pushnil(L);  /* Initial key. */
+
+  while(lua_next(L, -2)) {
+    /* We are only interested in String keys. */
+    if(lua_type(L, -2) == LUA_TSTRING) {
+      /* Avoid "_G" recursion in global table. The global table is also part of
+         global table :). */
+      if(!strcmp(lua_tostring(L, -2), "_G")) {
+        /* Remove value and continue. */
+        lua_pop(L, 1);
+        continue;
+      }
+
+      /* Is it the function we are looking for? */
+      if(lua_rawequal(L, fn_idx, -1)) {
+        /* Remove value and keep name. */
+        lua_pop(L, 1);
+        return true;
+      }
+      /* If not go one level deeper and get the value recursively. */
+      else if(findfield(L, fn_idx, level - 1)) {
+        /* Remove the table but keep name. */
+        lua_remove(L, -2);
+
+        /* Add a "." in between. */
+        lua_pushliteral(L, ".");
+        lua_insert(L, -2);
+
+        /* Concatenate last 3 values, resulting "table.some_func". */
+        lua_concat(L, 3);
+
+        return true;
+      }
+    }
+
+    /* Pop the value. */
+    lua_pop(L, 1);
+  }
+
+  return false;
+}
+
+
+/* Pushes a function name if found in the global table and returns true.
+   Returns false otherwise. */
+/* Expects the function to be pushed in the stack. */
+static bool pushglobalfuncname(lua_State *L) {
+  int top = lua_gettop(L);
+
+  /* Start from the global table. */
+  lua_pushglobaltable(L);
+
+  if(findfield(L, top, 2)) {
+    lua_remove(L, -2);
+    return true;
+  }
+
+  lua_pop(L, 1);
+  return false;
+}
+
+
+/* Returns the maximum number of levels in Lua stack. */
+static int countlevels(lua_State *L) {
+  lua_Debug ar;
+  int li = 1, le = 1;
+
+  /* Find an upper bound */
+  while (lua_getstack(L, le, &ar)) {
+    li = le, le *= 2;
+  }
+
+  /* Do a binary search */
+  while (li < le) {
+    int m = (li + le) / 2;
+
+    if (lua_getstack(L, m, &ar)) li = m + 1;
+    else le = m;
+  }
+
+  return le - 1;
+}
+
+
+/* Counts the number of white and black frames in the Pallene call stack. */
+static void countframes(pt_fnstack_t *fnstack, int *mwhite, int *mblack) {
+  *mwhite = *mblack = 0;
+
+  for(int i = 0; i < fnstack->count; i++) {
+    *mwhite += (fnstack->stack[i].type == PALLENE_TRACER_FRAME_TYPE_C);
+    *mblack += (fnstack->stack[i].type == PALLENE_TRACER_FRAME_TYPE_LUA);
+  }
+}
+
+
+/* Responsible for printing and controlling some of the traceback fn parameters. */
+static void makeitcount(lua_State *L, luaL_Buffer *buf, bool *ellipsis, int *pframes, int nframes) {
```

Comment:
This function needs a more descriptive name & comment.

Two of the arguments are mutable state, passed via pointers. Can we avoid that or at least reduce the amount of mutable state?

1. I think we could increment the pframes outside of makeitcount. Yes it's "code duplication" but I think it will be easier to understand than a pointer.
2. We can deduce ellipsis boolean similarly to what we did for should_print. If the frame number is exactly equal to the smallest integer that would not be printed, then we print the ellipsis message.
3. Give longer names to pframes and nframes, because both are "number of frames". It sounds like pframe is current frame and nframe is total frames?

Reply:
1. Sure.
2. Actually we can't. The deduction part is easy and obvious, but how would we know have we printed the ellipsis message? Introducing another state for that is waste and even if we do, we have to pass it to makeitcount regardless. By the way, `makeitcount` is now `render`, which makes more sense in this case.
3. Correct. pframes = Printed frames or number of frames. nframes is self explanatory.

Reply:
Got your point, fixed.